chalice/app.py

```diff
@@ -688,6 +688,10 @@ def __init__(self, app_name, debug=False, configure_logs=True, env=None):
         if env is None:
             env = os.environ
         self._initialize(env)
+        self.experimental_feature_flags = set()
+        # This is marked as internal but is intended to be used by
+        # any code within Chalice.
+        self._features_used = set()
```

Comment:
Hmm I'm not sure if I like this trend of marking something as an internal property and then accessing it in the deployer code. I think we did something similar in blueprints. I get why we are doing it right now, but I'm wondering if there are any alternatives to tell users that it is an internal property? Options that I am thinking include: Just documenting it as internal (somewhat similar to some of the standard library) or having like an property like `app.internal_stuff` that could be a grab bag for anything that needs to be accessed in the deployer but should not be used in a user's Chalice app.

Reply:
I think we've learned that lesson from botocore that this isn't a great idea.  It basically comes down to either:

a) We make it public, call it something that's hopefully obvious (e.g. `internal_stuff`) and document that users shouldn't touch it and hope that they don't.

b) Make it actually internal with a `_` prefix, and then we (chalice devs) purposefully access internal attributes we know are safe.

I think (b) is a much safer alternative, especially given that the `_` prefix is a much more explicit mechanism.   We need the concept of "package private" and I would prefer the option that requires chalice devs to be mindful of what they're accessing (e.g that `_features_used` is safe) vs. customers knowing that they shouldn't access an attribute called `internal_features_used`.

Reply:
Agree with james. I think the convention for a friend-ish class should just be marked private, its a lot safer for everyone involved.

Reply:
I guess its fine. The only thing I do not like about the `_` prefix is that it can be hard to know which `_` properties are public for our development only. The only way to really tell is to look for the comment in the code where it is defined. We also have to be diligent in making sure we test the `_` property so we do not accidentally refactor it in a backwards incompatible way (but I think we have a good amount of tests for the ones we have added so that should be fine).